### PR TITLE
feat: Zigbee Specification revamp

### PR DESCRIPTION
**Pull request by Nerivec**
- Move zcl into zspec folder.
- Add first draft for ZDO specification [R23]. _This PR doesn't introduce its usage, it needs extensive testing/tweaking before it can be refactored in._
- Add ZCL-specific test (independent of higher-level herdsman logic).
- Reorganized imports/exports for zspec
  - consts/enums will now be automatically exported whenever new ones are added in their respective places.
  - `import {ZSpec, Zcl, Zdo} from 'zigbee-herdsman'`
  - Zigbee consts => `ZSpec.ZIGBEE_REVISION`
  - Zigbee enums => `ZSpec.BroadcastAddress.RX_ON_WHEN_IDLE`
  - Zigbee util functions => `ZSpec.Utils.channelsToUInt32Mask()`
  - ZDO consts => `Zdo.ZDO_ENDPOINT`
  - ZDO enums => `Zdo.LeaveRequestFlags.WITHOUT_REJOIN`
  - ZDO util functions => `Zdo.Utils.getServerMask()`
  - ZDO status/error => `Zdo.Status.SUCCESS` and `new Zdo.StatusError(Zdo.Status.INV_REQUESTTYPE)`
  - ZCL consts => `Zcl.POWER_SOURCES`
  - ZCL enums => `Zcl.DataType.UINT8`
  - ZCL util functions => `Zcl.Utils.isClusterName()`
  - ZCL status/error => `Zcl.Status.SUCCESS` and `new Zcl.StatusError(Zcl.Status.FAILURE)`
  - ZCL frame => `Zcl.Frame`
  - ZCL header => `Zcl.Header`
- Add enums: `ParameterCondition`, `DataTypeClass` and replace hardcoded strings in Clusters/Foundation.
- Fix typing for `ParameterDefinition`, and more...
- Extract Xiaomi workaround from `CHAR_STR` to its own data type `MI_STRUCT`.
- Add several basic consts that should eventually replace hardcoded stuff throughout the code. _Can be swapped when encountered in future PRs. More can be added as needed._
- Added several basic Zigbee types as "more informative" wrappers for Typescript types. _Same as above._

TODO:
- [x] ZDO buffalo tests
- [ ] Update zigbee2mqtt imports
- [x] https://github.com/Koenkk/zigbee-herdsman-converters/pull/7488
- [ ] Update zigbee2mqtt-frontend imports
- [ ] Remove old tests now covered by new ZCL-specific ones (_kept unchanged until merge to ensure no change in logic creeps up_)

### CODE REVIEW

**Comment by Nerivec** [comment]
@Koenkk If you can review [this commit](https://github.com/Koenkk/zigbee-herdsman/pull/1042/commits/93346068fca2ed3d76362e7cab4d0d9852e9ec02) in great details. It drastically changes the code for clusters/custom clusters lookup. We need to make sure it still matches the way you designed custom clusters to be (i.e. that tests aren't testing this wrong). 😅
If it does, these changes have one heck of a nice side-effect:

x10000 `getCluster`:
Old: ~280ms
New: ~50ms

x10000 `getCluster` with 2 custom clusters:
Old: ~7.5s
New: ~57ms

File: src/zspec/zcl/utils.ts [hunk 1]
```diff
@@ -32,37 +31,73 @@ export function getDataTypeClass(dataType: DataType): DataTypeClass {
...
         }
     } else {
         name = key;
     }
 
-    let cluster = clusters[name];
+    let cluster = (name != undefined) && hasCustomClusters(customClusters) ?
+        {
+            ...Clusters[name as ClusterName],
+            ...customClusters[name],// should override Zcl clusters
```
Comment (Koenkk) [inline src/zspec/zcl/utils.ts:98]: I think this old logic is missing here:

```
        ...customClusters, // Custom clusters have a higher priority than Zcl clusters (custom cluster has a DIFFERENT name than Zcl clusters)
        ...Clusters,
        ...customClusters, // Custom clusters should override Zcl clusters (custom cluster has the SAME name than Zcl clusters)
```

Besides that, this commit looks OK

Comment (Nerivec) [reply src/zspec/zcl/utils.ts:98]: The same logic should be applied in the order the lookup is now done (looks into custom clusters first).
Current tests are passing, but if you want to add a couple very specific in the way you intend this to work?

Comment (Koenkk) [reply src/zspec/zcl/utils.ts:98]: It's not tested by tests yet, but it would happen if you define a cluster with an already existing ID.

Comment (Nerivec) [reply src/zspec/zcl/utils.ts:98]: Isn't this one doing it?
https://github.com/Koenkk/zigbee-herdsman/blob/f1890adb3678c86f209122b0cc7ddb6f7eb8a497/test/controller.test.ts#L2438

Comment (Koenkk) [reply src/zspec/zcl/utils.ts:98]: This tests defines a attribute for an existing cluster, what I mean is defining a new cluster of which the ID is already used by an existing cluster.

Comment (Nerivec) [reply src/zspec/zcl/utils.ts:98]: This test is the same as the one linked above, except it should override the ZCL cluster. (`name` is not ZCL, but ID is.)
Can you confirm that is correct?
```typescript
    it('Should allow to specific custom cluster as override for Zcl cluster', async () => {
        await controller.start();
        await mockAdapterEvents['deviceJoined']({networkAddress: 129, ieeeAddr: '0x129'});
        const device = controller.getDeviceByIeeeAddr('0x129');
        device.addCustomCluster('myCustomCluster', {ID: Zcl.Clusters.genBasic.ID, commands: {}, commandsResponse: {}, attributes: {customAttr: {ID: 256, type: Zcl.DataType.UINT8}}});
        const buffer = Buffer.from([24,169,10,0,1,24,3,0,0,24,1]);
        const header = Zcl.Header.fromBuffer(buffer);
        await mockAdapterEvents['zclPayload']({wasBroadcast: false, address: 129, clusterID: Zcl.Clusters.genBasic.ID, data: buffer, header, endpoint: 1, linkquality: 50, groupID: 1});
        expect(events.message.length).toBe(1);
        expect(events.message[0].data).toStrictEqual({customAttr: 3, 0: 1/*zclVersion no longer recognized, cluster is overridden*/});
        expect(events.message[0].cluster).toBe('myCustomCluster');
    });
```

Comment (Koenkk) [reply src/zspec/zcl/utils.ts:98]: I updated the test in https://github.com/Koenkk/zigbee-herdsman/pull/1042/commits/a97271bd1a798d9775bde591f8ffe9a1d1195d3b and it passes, looks good!

Comment (Nerivec) [reply src/zspec/zcl/utils.ts:98]: That's just attribute overrides in your commit though. Should I add the test mentioned above for complete cluster override too ("same cluster ID, different name", results in custom cluster used)?

Comment (Koenkk) [reply src/zspec/zcl/utils.ts:98]: Yes please!

**Comment by Koenkk** [comment]
Thanks!

**Pull request merged**